### PR TITLE
GEOMESA-1438 Support for json attributes in SimpleFeatures

### PR DESCRIPTION
**Pull request by elahrvivaz**
* Attributes can be marked as json in the simple feature type definition: 'json-field:String:json=true'
* Json attributes will be serialized as bson
* Nested json can be queried or projected using json-path (note json path must be double-quoted): "$.json-field.foo" = 'bar'
* Removed old serialization code - dropping support for versions prior to 1.1.0-rc.1

Signed-off-by: Emilio Lahr-Vivaz <elahrvivaz@ccri.com>

### CODE REVIEW

**Comment by jnh5y** [comment]
@elahrvivaz How much of this is the code/changes for "Removed old serialization code - dropping support for versions prior to 1.1.0-rc.1"? 

I think that might be worth a separate PR.  Also, we can likely merge that more quickly.  If we want to do that, it should be the 1.3.0 milestones ASAP.

File: build/cqs.tsv [hunk 1]
```diff
@@ -148,6 +148,8 @@ org.locationtech.sfcurve:sfcurve-api_2.11	0.2.0	compile
 org.locationtech.sfcurve:sfcurve-zorder_2.11	0.2.0	compile
 org.objenesis:objenesis	1.0	compile
 org.ow2.asm:asm	4.0	compile
+org.parboiled:parboiled-core	1.1.7	compile
+org.parboiled:parboiled-scala_2.11	1.1.7	compile
```
Comment (elahrvivaz) [inline build/cqs.tsv]: https://dev.eclipse.org/ipzilla/show_bug.cgi?id=11990

**Comment by elahrvivaz** [comment]
@jnh5y basically all the deleted files are from removing the old code. I'm not sure if it is worth being a separate PR though - the main impetus was to not have to worry about handling json serialization in the old framework. Seems pointless to remove functionality if the added functionality isn't getting merged yet.

File: geomesa-features/geomesa-feature-kryo/src/main/scala/org/locationtech/geomesa/features/kryo/json/JsonPathPropertyAccessor.scala [hunk 2]
```diff
@@ -0,0 +1,122 @@
...
+  * Access values from a json-type string field. Syntax must start with '$.'. The first part of the path
+  * selects the simple feature attribute, and the rest of the path selects within the json contained in
+  * that attribute.
+  *
+  * Note: this class is optimized for `KryoBufferSimpleFeature`s. It will work on standard simple features,
+  * but will incur a serialization cost.
+  */
+object JsonPathPropertyAccessor extends PropertyAccessor {
+
+  // TODO once jayway CQs are approved for GEOMESA-1406, we could change the non-kryo implementation to use that
```
Comment (jnh5y) [inline geomesa-features/geomesa-feature-kryo/src/main/scala/org/locationtech/geomesa/features/kryo/json/JsonPathPropertyAccessor.scala]: Lemme ping the PMC about the CQs for this one and the jayway bits.

**Comment by jnh5y** [comment]
@mdzimmerman @cfkelly can you read through this PR?

File: geomesa-features/geomesa-feature-kryo/src/main/scala/org/locationtech/geomesa/features/kryo/json/JsonPathParser.scala [hunk 3]
```diff
@@ -0,0 +1,113 @@
...
+
+  def Character: Rule0 = rule { EscapedChar | NormalChar }
+
+  def EscapedChar: Rule0 = rule { "\\" ~ (anyOf("\"\\/bfnrt") | Unicode) }
+
+  def NormalChar: Rule0 = rule { "a" - "z" | "A" - "Z" | "0" - "9" }
+
+  def Unicode: Rule0 = rule { "u" ~ HexDigit ~ HexDigit ~ HexDigit ~ HexDigit }
+
+  def HexDigit: Rule0 = rule { "0" - "9" | "a" - "f" | "A" - "Z" }
```
Comment (cfkelly) [inline geomesa-features/geomesa-feature-kryo/src/main/scala/org/locationtech/geomesa/features/kryo/json/JsonPathParser.scala]: Shouldn't the third one be "A" - "F"?

Comment (elahrvivaz) [inline geomesa-features/geomesa-feature-kryo/src/main/scala/org/locationtech/geomesa/features/kryo/json/JsonPathParser.scala]: yeah, good catch

Comment (elahrvivaz) [inline geomesa-features/geomesa-feature-kryo/src/main/scala/org/locationtech/geomesa/features/kryo/json/JsonPathParser.scala]: fixed

File: geomesa-features/geomesa-feature-kryo/src/test/scala/org/locationtech/geomesa/features/kryo/json/JsonPathPropertyAccessorTest.scala [hunk 4]
```diff
@@ -0,0 +1,61 @@
...
+  "JsonPathPropertyAccessor" should {
+    "be available on the classpath" in {
+      import scala.collection.JavaConverters._
+      val accessors =
+        PropertyAccessors.findPropertyAccessors(new ScalaSimpleFeature("", sft), "$.json.foo", classOf[String], null)
+      accessors must not(beNull)
+      accessors.asScala must contain(JsonPathPropertyAccessor)
+    }
+    "access json values in simple features" in {
+      val property = ff.property("$.json.foo")
```
Comment (elahrvivaz) [inline geomesa-features/geomesa-feature-kryo/src/test/scala/org/locationtech/geomesa/features/kryo/json/JsonPathPropertyAccessorTest.scala]: added

File: geomesa-features/geomesa-feature-kryo/src/main/scala/org/locationtech/geomesa/features/kryo/json/KryoJsonSerialization.scala [hunk 5]
```diff
@@ -0,0 +1,522 @@
...
+  private def writeDocument(out: Output, name: String, value: JObject): Unit = {
+    out.writeByte(DocByte)
+    out.writeName(name)
+    writeDocument(out, value)
+  }
+
+  // write a document without a name - used for the outer-most object which doesn't have a key
+  private def writeDocument(out: Output, value: JObject): Unit = {
+    val start = out.position()
+    out.setPosition(start + 4) // skip space to write total length
```
Comment (elahrvivaz) [inline geomesa-features/geomesa-feature-kryo/src/main/scala/org/locationtech/geomesa/features/kryo/json/KryoJsonSerialization.scala]: updated

File: geomesa-utils/src/main/scala/org/locationtech/geomesa/utils/geotools/SimpleFeatureTypes.scala [hunk 6]
```diff
@@ -256,9 +258,12 @@ object SimpleFeatureTypes {
       if (ad.isBinTrackId) {
         options.put(OPT_BIN_TRACK_ID, "true")
       }
-      if (ad.isKeepStats) {
+      if (ad.isKeepStats()) {
```
Comment (jahhulbert-ccri) [inline geomesa-utils/src/main/scala/org/locationtech/geomesa/utils/geotools/SimpleFeatureTypes.scala]: picky...remove () maybe? or add to isBintrackid?

File: geomesa-features/geomesa-feature-kryo/src/main/scala/org/locationtech/geomesa/features/kryo/json/KryoJsonSerialization.scala [hunk 7]
```diff
@@ -0,0 +1,522 @@
...
+import org.json4s.native.JsonMethods.{parse => _, _}
+import org.locationtech.geomesa.features.kryo.json.JsonPathParser.JsonPathFunction.JsonPathFunction
+import org.locationtech.geomesa.features.kryo.json.JsonPathParser._
+
+import scala.util.control.NonFatal
+
+/**
+  * Serializes into bson (http://bsonspec.org/). Note this is a limited form of bson that only matches
+  * the existing json types - does not cover the bson extensions like binary data, etc. Also note endianness,
+  * etc might not match the spec 100%.
```
Comment (jahhulbert-ccri) [inline geomesa-features/geomesa-feature-kryo/src/main/scala/org/locationtech/geomesa/features/kryo/json/KryoJsonSerialization.scala]: how do I run GeoRaspberry PiMesa?

**Comment by jahhulbert-ccri** [comment]
looks pretty good. is there a follow on ticket for docs?

File: geomesa-utils/src/main/scala/org/locationtech/geomesa/utils/geotools/SimpleFeatureTypes.scala (hunk 6 above)
Comment (elahrvivaz) [inline geomesa-utils/src/main/scala/org/locationtech/geomesa/utils/geotools/SimpleFeatureTypes.scala]: yeah, i prefer to have the `()` to indicate it's not a straight retrieval but it's actually doing a little processing. not sure if that's really a standard :P

**Comment by elahrvivaz** [comment]
yeah, docs ticket here: https://geomesa.atlassian.net/browse/GEOMESA-1457

File: geomesa-utils/src/main/scala/org/locationtech/geomesa/utils/geotools/SimpleFeatureTypes.scala (hunk 6 above)
Comment (elahrvivaz) [inline geomesa-utils/src/main/scala/org/locationtech/geomesa/utils/geotools/SimpleFeatureTypes.scala]: fixed

**Pull request merged**